### PR TITLE
support different type in tensorflow loader

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

support different type in tensorflow loader

## How was this patch tested?

unit tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 1]
```diff
@@ -348,4 +349,49 @@ object Utils {
     }
     (padH, padH, padW, padW, oheight, owidth)
   }
+
+  private[nn] def allocate(shape: Array[Int], dataType: TensorDataType): Tensor[_] = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: move this to Activity?

Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: Move this to Tensor?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala [hunk 2]
```diff
@@ -348,4 +349,49 @@ object Utils {
...
+      case CharType =>
+        Tensor[Char](shape)
+      case StringType =>
+        Tensor[ByteString](shape)
+      case BooleanType =>
+        Tensor[Boolean](shape)
+    }
+  }
+
+  private[nn] def copy(dest: Tensor[_], src: Tensor[_]): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Utils.scala]: revise tensor.copy to support this?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/JoinTable.scala [hunk 3]
```diff
@@ -109,29 +116,29 @@ class JoinTable[T: ClassTag] (
...
     while (i < input.length) {
-      val currentOutput = input(i + 1).asInstanceOf[Tensor[T]]
+      val currentOutput = input(i + 1).asInstanceOf[Tensor[_]]
       val _offset = offset
       val _i = i
       results(i) = Engine.model.invoke( () => {
         val narrowedTensor = gradOutput.narrow(dimension, _offset, currentOutput.size(dimension))
-        if (!gradInput.contains(_i + 1)) gradInput(_i + 1) = Tensor()
-        gradInput[Tensor[T]](_i + 1).resizeAs(input(_i + 1))
+        if (!gradInput.contains(_i + 1)) gradInput(_i + 1) =
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/JoinTable.scala]: revise resizeAs?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Const.scala [hunk 4]
```diff
@@ -29,8 +29,8 @@ private[bigdl] trait WithoutInput
  * @param value the constant tensor to be returned in forward
  */
 @SerialVersionUID(-4008935551091949324L)
-private[bigdl] class Const[T: ClassTag](value: Tensor[T])(implicit ev: TensorNumeric[T])
-  extends AbstractModule[Activity, Tensor[T], T] with WithoutInput {
+private[bigdl] class Const[T: ClassTag](value: Tensor[_])(implicit ev: TensorNumeric[T])
+  extends AbstractModule[Activity, Activity, T] with WithoutInput {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Const.scala]: Why not Tensor[B]?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala [hunk 5]
```diff
@@ -15,38 +15,96 @@
...
 
 /**
  * Creates a tensor filled with a scalar value. Input should be a 1-D tensor defining
  * the shape of the output tensor.
- * @param value the scalar value to be filled.
  */
 @SerialVersionUID(-471757174144422555L)
-private[bigdl] class Fill[T: ClassTag](value: T) (implicit ev: TensorNumeric[T])
-  extends TensorModule[T] {
+private[bigdl] class Fill[T: ClassTag]() (implicit ev: TensorNumeric[T])
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala]: value B?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala [hunk 6]
```diff
@@ -15,38 +15,96 @@
...
+  extends AbstractModule[Table, Tensor[_], T] {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala]: [Tensor[B], Tenosr[B], T]

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala (hunk 5 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala]: B is inferenced from input, we do not know B in advance

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala (hunk 6 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala]: The same as above

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala [hunk 7]
```diff
@@ -15,38 +15,98 @@
...
+
     output
   }
 
-  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
-    gradInput.resizeAs(input).zero()
+  private def fill(output: Tensor[_], valueTensor: Tensor[_]): Unit = {
+    require(output.getType() == valueTensor.getType(),
+      "fill tensor and value tensor are not the same type")
+    valueTensor.getType() match {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/tf/Fill.scala]: revise fill to handle this?

**Comment by yangw1234** [comment]
pr validation pass

**Pull request merged**